### DISCUSSION TITLE
Adding surrounds for LaTeX

### DISCUSSION THREAD

**Discussion by benbrastmckie** [kylechui/nvim-surround#175] [Q&A]
I would like to be able to surround a text object `x` with `\textit{   }`, and similarly for other LaTeX operations. I'm struggling to achieve this by changing the default config that is provided. Any help would be much appreciated.

**Comment by kylechui** [comment]
Hi there! You can try something like this:
```lua
require("nvim-surround").setup({
    surrounds = {
        ["x"] = {
            add = { "\\textit{", "}" },
            find = "\\textit%b{}",
            delete = "^(\\textit{)().-(})()$",
        },
    }
})
```

For an explanation on how/why this works, please see `:h nvim-surround.config.surrounds`. If there's anything that's unclear, please let me know! The `find` and `delete` keys are Lua patterns, where `%b{}` indicates matching a balanced set of curly braces. The weird parentheses in `delete` are Lua match groups, explained more in `:h nvim-surround.config.get_selections()`.

Bonus: [This](https://github.com/kylechui/nvim-surround/discussions/53#discussioncomment-3142459) might be of some interest to you if you're working with LaTeX commands a bit, as a generalization of the surround you're asking for.

**Reply by benbrastmckie** [reply]
Thanks so much! That worked great. Few other related questions...

(1) I've included and begun to edit the config file provided in the repo, requiring this in my init.lua. Is this necessary? Or can I just include parts of that config file that I want to change?

(2) I have noticed that `b` still triggers parentheses to be added, and I'd like to remove this, as `b` means boldface for my purposes.

(3) More generally, I'd like to turn off all default mappings if possible, and then add the ones I need. This is how I used to do it with the old vimscript surround plugins. Is this possible? Or do I need `= false` following each of the default mappings defined in the config file?

Thanks again for your help!

Ben

**Reply by kylechui** [reply]
1. You only need to include the portions that you wish to modify, *not* the entire default configuration.
2. You can disable this in the `aliases` key, by setting `["b"] = false`.
3. There is currently no option to disable all keymaps in one action (sorry!). You would need to set each one yourself, i.e.
```lua
require("nvim-surround").setup({ -- Re-enable the ones you wish to use!
    keymaps = {
        insert = false,
        insert_line = false,
        normal = false,
        normal_cur = false,
        normal_line = false,
        normal_cur_line = false,
        visual = false,
        visual_line = false,
        delete = false,
        change = false,
    },
})
```

A few additional notes/suggestions based on the information you've provided:
1. It might be in your interest to copy the default configuration in its entirety and go through each element one by one, and disable it if you don't think it's strictly necessary. If you have any questions about what certain things do, please consult the help menu or leave another message here and I can see what I can do to help.
2. There does exist a function `require("nvim-surround").buffer_setup`. It performs identically to `setup`, but essentially "overrides" your "global configuration" for the current buffer. I mention this because you can use this to setup filetype-specific surrounds, e.g. by putting this in `ftplugin/tex.lua` of your config. For examples, see [my config here](https://github.com/kylechui/config.nvim/blob/main/after/ftplugin/tex.lua). This would be useful if you wish to "overload" the usage of `b` for different filetypes, i.e. bolding using `\textbf` in LaTeX files, bolding using `**` in Markdown files, etc.

Bonus: I haven't personally done this yet, but it is theoretically possible for one to use `b` to add `\textbf` to the buffer in text mode and `\mathbf` in math mode by running a custom function in the surround. If you're interested in this, I can take some time to figure it out! :slightly_smiling_face:

**Reply by benbrastmckie** [reply]
Thanks so much! It is working great :)

And yes, it would be great to be able to switch to \mathbf in math mode.

Great plugin, thanks again!

**Reply by kylechui** [reply]
This seems like it would do the trick:
```lua
-- Put this in ftplugin/tex.lua, for example
require("nvim-surround").buffer_setup({
    surrounds = {
        ["b"] = {
            add = function()
                if vim.fn["vimtex#syntax#in_mathzone"]() == 1 then
                    return { { "\\mathbf{" }, { "}" } }
                end
                return { { "\\textbf{" }, { "}" } }
            end,
            find = "\\%a-bf%b{}",
            delete = "^(\\%a-bf{)().-(})()$",
        },
    },
    aliases = {
        b = false,
    },
})
```

Also as an aside, I just wanted to say thanks for making your YouTube videos; I remember watching them 1-2 years ago when I was first getting started with Neovim :sparkling_heart:

**Reply by benbrastmckie** [reply]
Thanks so much! I've been working on switching to Lua, after which I'm going to make a new video series which will hopefully be a little more streamlined than the previous videos. As I'm sure you can tell, I'm new to Lua, but excited to keep learning, and want to help others coming from LaTeX (and with no knowledge of the terminal etc.) how to learn to use these great tools.

(A) I tried the code you included above, but with no luck. It does wrap text in the non-math-mode command `\textbf{ }`, but I can't get it to spit our `\mathbf{ }` in context such as:
```
$ \Sigma $ --> $ \mathit{\Sigma} $

\[ 

  \Sigma --> \mathtt{\Sigma}

\]
```

(B) One further question... If I want to wrap 'part of a line' with `$` resulting in `$ part of a line $`, I have the following:
```
local M = {}

local config = require("nvim-surround.config")

require("nvim-surround").buffer_setup({
  surrounds = {
    ["$"] = {
      add = { "$ ", " $" },
      find = function()
        return M.get_selection({ motion = "a$" })
      end,
      delete = "^($)().-($)()$",
    },

return M
```

I'm not sure if the `M` is required, or if there is a better way to do this. 

(C) Also, is there a way to insert `\[ ... \]` separated by lines as follows, reusing `$` to trigger the surround:
```
\[

first math line
second math line

\]
```
I'm glad that you liked the series on LaTeX in NeoVim! If you had any suggestions for how I could improve the Lua [config](https://github.com/benbrastmckie/.config/tree/lua) I'm working on, I'd be very curious to know. Unfortunately I'm still struggling to get the LSP to import from Zotero.bib in `~/texmf/bibtex/bib/` by including `\bibliography{Zotero.bib}` at the end of the document, but so it goes...

**Reply by kylechui** [reply]
A) That's a bit strange; I assume you mean `bf` as opposed to `it` and `tt`? I just tried it again and it seems to work on my end.
B) A few things:
* The `local M` stuff is actually only necessary on my end for the plugin; it's how I "export" my functions as a "module" for you to use. From your end, you only need the `buffer_setup` function and `local config` line, and `M.get_selection` should be `config.get_selection`.
* Unlike `vim-surround`, `nvim-surround`'s default action on "undefined" characters is to treat the singular characters as a left/right pair, for surrounding and deleting. Thus you actually don't need to do any additional setup, and can just select text and do `<S-s>$` out of the box (I took a look at your setup).

C) There is currently no way of "overloading" the behavior so the `$` key performs multiple actions; one thing that you could do is to use a different key like `e` (for equation) instead. 

As for the Zotero stuff; I unfortunately never got far enough in math to utilize a tool like Zotero for citations, so I don't know how to help with that. I changed majors to computer science after my second year :sweat_smile: 

I'll take a look at your config!

**Reply by kylechui** [reply]
I opened [a PR](https://github.com/benbrastmckie/.config/pull/88) on your config with a few comments; I hope you find it useful :)

**Reply by benbrastmckie** [reply]
(A) Yes, sorry, I have it set up for all the text functions I use like `\texttt{}` etc., and cut and paste the wrong ones. I still can't seem to get it to work in mathmode, but otherwise works great. I'll keep fussing with it and hopefully get it working.

(B)i Awesome glad to sort that out!

(B)ii The only reason I define something for `$` is that I like spaces between the `$` and the math I'm selecting. Not a big deal really since it is rare to use this surround anyhow, and autopairs gives me the extra spaces I'm looking for upon inserting new `$ | $`.

(C) Ok, that's no big deal. Just thought I'd check :)

I also made my way from math (undergrad) to logic (grad). Now starting to make my way into compsci a bit. We will see where it leads...

Thanks again for the help!

**Reply by benbrastmckie** [reply]
And regarding the PR... I messed up the merge so had to force push, but thanks so much. The changes were all made :)
